File: src/domain/task/TaskExecutionHistoryService.ts
import { EntityManager } from "typeorm";
import { ExecutionStatus } from "../../contracts/enums";
import { NotFoundError } from "../../errors/AppError";
import { TaskExecutionHistoryEntity } from "./TaskExecutionHistory.entity";
import { TaskExecutionHistoryRepository } from "./TaskExecutionHistoryRepository";
import { TaskRepository } from "./TaskRepository";

/**
 * TaskExecutionHistoryService – Execution attempt tracking and lifecycle.
 * Each task rerun creates a new execution history record with incremented attempt_number.
 * Snapshots task input at execution time.
 */
export class TaskExecutionHistoryService {
  constructor(
    private readonly executionHistoryRepo: TaskExecutionHistoryRepository,
    private readonly taskRepo: TaskRepository
  ) {}

  /**
   * Create a new execution history record for a task.
   * - Gets the next attempt_number (max + 1)
   * - Snapshots the current task input
   * - Sets executionStatus = Running
   * - Updates Task.latestExecutionId
   */
  async createExecution(taskId: string, em?: EntityManager): Promise<TaskExecutionHistoryEntity> {
    const task = await this.taskRepo.findById(taskId, em);
    if (!task) throw new NotFoundError("Task", taskId);

    // Determine next attempt number
    const maxAttempt = await this.executionHistoryRepo.getMaxAttemptNumber(taskId, em);
    const nextAttempt = maxAttempt + 1;

    // Create execution record
    const execution = new TaskExecutionHistoryEntity();
    execution.taskId = taskId;
    execution.attemptNumber = nextAttempt;
    execution.executionStatus = "Running";
    execution.inputSnapshotJson = task.inputParametersJson;
    execution.resultSummaryJson = null;
    execution.resultLogs = null;
    execution.startTime = new Date();
    execution.endTime = null;

    const saved = await this.executionHistoryRepo.save(execution, em);

    // Update Task.latestExecutionId
    await this.taskRepo.updateTask(task, { latestExecutionId: saved.id }, em);

    return saved;
  }

  /**
   * Retrieve all execution attempts for a task, ordered by attempt_number.
   */
  async findByTaskId(taskId: string, em?: EntityManager): Promise<TaskExecutionHistoryEntity[]> {
    return this.executionHistoryRepo.findByTaskId(taskId, em);
  }

  /**
   * Retrieve the latest execution attempt for a task.
   */
  async findLatest(taskId: string, em?: EntityManager): Promise<TaskExecutionHistoryEntity | null> {
    return this.executionHistoryRepo.findLatestByTaskId(taskId, em);
  }

  /**
   * Mark an execution as complete with result.
   * Sets executionStatus, resultSummaryJson, resultLogs, and endTime.
   */
  async completeExecution(
    executionId: string,
    executionStatus: ExecutionStatus,
    resultSummaryJson?: string,
    resultLogs?: string,
    em?: EntityManager
  ): Promise<TaskExecutionHistoryEntity> {
    const execution = await this.executionHistoryRepo.findById(executionId, em);
    if (!execution) throw new NotFoundError("TaskExecutionHistory", executionId);

    execution.executionStatus = executionStatus;
    if (resultSummaryJson !== undefined) execution.resultSummaryJson = resultSummaryJson;
    if (resultLogs !== undefined) execution.resultLogs = resultLogs;
    execution.endTime = new Date();

    return this.executionHistoryRepo.save(execution, em);
  }
}

File: src/main/java/com/wwa/agenthub/domain/task/TaskExecutionHistoryService.java
package com.wwa.agenthub.domain.task;

import com.wwa.agenthub.contracts.enums.ExecutionStatus;
import com.wwa.agenthub.errors.NotFoundAppException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Clock;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * TaskExecutionHistoryService – Execution attempt tracking and lifecycle.
 * Each task rerun creates a new execution history record with an incremented attemptNumber.
 * The current task input is snapshotted at execution time.
 */
@Service
@RequiredArgsConstructor
public class TaskExecutionHistoryService {

    private final TaskExecutionHistoryRepository executionHistoryRepository;
    private final TaskRepository taskRepository;
    private final TaskService taskService;
    private final Clock clock;

    /**
     * Create a new execution history record for a task.
     * <ul>
     *   <li>Gets the next attemptNumber (max + 1)</li>
     *   <li>Snapshots the current task inputParameters</li>
     *   <li>Sets executionStatus = Running</li>
     *   <li>Updates Task.latestExecutionId</li>
     * </ul>
     */
    @Transactional
    public TaskExecutionHistory createExecution(String taskId) {
        Task task = taskRepository.findById(taskId)
                .orElseThrow(() -> new NotFoundAppException("Task", taskId));
        taskService.assertTaskRequestActive(task);

        int maxAttempt = executionHistoryRepository.findMaxAttemptNumberByTaskId(taskId);
        int nextAttempt = maxAttempt + 1;

        TaskExecutionHistory execution = new TaskExecutionHistory();
        execution.setTask(task);
        execution.setAttemptNumber(nextAttempt);
        execution.setExecutionStatus(ExecutionStatus.Running);
        execution.setInputSnapshot(task.getInputParameters());
        execution.setResultSummary(null);
        execution.setResultLogs(null);
        execution.setStartTime(clock.instant());
        execution.setEndTime(null);

        TaskExecutionHistory saved = executionHistoryRepository.save(execution);

        task.setLatestExecutionId(saved.getId());
        taskRepository.save(task);

        return saved;
    }

    /** Retrieve all execution attempts for a task, ordered by attemptNumber ascending. */
    @Transactional(readOnly = true)
    public List<TaskExecutionHistory> findByTaskId(String taskId) {
        return executionHistoryRepository.findByTaskIdOrderByAttemptNumberAsc(taskId);
    }

    /** Retrieve the latest execution attempt for a task. */
    @Transactional(readOnly = true)
    public Optional<TaskExecutionHistory> findLatest(String taskId) {
        return executionHistoryRepository.findFirstByTaskIdOrderByAttemptNumberDesc(taskId);
    }

    /**
     * Mark an execution attempt as complete with result data.
     * Sets executionStatus, resultSummary, resultLogs, and endTime.
     */
    @Transactional
    public TaskExecutionHistory completeExecution(String executionId,
                                                  ExecutionStatus executionStatus,
                                                  Map<String, Object> resultSummary,
                                                  String resultLogs) {
        TaskExecutionHistory execution = executionHistoryRepository.findById(executionId)
                .orElseThrow(() -> new NotFoundAppException("TaskExecutionHistory", executionId));

        execution.setExecutionStatus(executionStatus);
        if (resultSummary != null) execution.setResultSummary(resultSummary);
        if (resultLogs != null) execution.setResultLogs(resultLogs);
        execution.setEndTime(clock.instant());

        return executionHistoryRepository.save(execution);
    }
}
